proposals/0000-local-import.rst

```diff
@@ -0,0 +1,225 @@
+   
+The following changes in the Haskell 2010 grammar are required:
+
+::
+
+  decls → { impdecls ; decl_1 ; … ; decl_n }    (n ≥ 0)
+        | { decl_1 ; … ; decl_n }
+  lexp  → …
+        | modid.{ lexp }
+        | do { stmts }
+        | do { impdecls ; stmts }
+
+
+Effect and Interactions
+-----------------------
+This proposal strictly extends the language, without affecting the behavior of existing code. It does not interact with any existing language extension.
+
+The changes give programmers various ways to reduce the number of toplevel imports, to limit their effect to specifics parts of the code and to convey intent about their uses. The shortcut syntax can be especially useful for scoping module imports over expressions with operators in DSLs.
+
+Furthermore, it encourages the use of qualified imports, as those can be locally “de-qualified” in order to improve readability.
+
+Costs and Drawbacks
+-------------------
+These changes should be easy to grasp by beginners. Moreover, they could greatly improve learnability of Haskell libraries, by expliciting the module each function comes from in examples and tutorials.
+
+Some existing tooling (e.g., ``snack``) assume that imports are only found at the toplevel and might be broken by this change. Perhaps more importantly, this change would make it harder for IDE-like tools such as ``hie`` to determine the set of valid completions ; such tools would need to be made context-sensitive, like OCaml's merlin.
+
+This change would make it harder to determine at first glance inter-dependencies between modules.
+
+Finally, some library writers might choose to design their library around this extension. Using such libraries without this extension enabled might be inconvenient, which could be perceived as a drawback by some users.
+
+Alternatives
+------------
+To our knowledge, there is no other language feature or extension providing similar benefits.
+
+It might be valuable to also allow some form of typelevel local import ; the shortcut syntax, in particular, could be used to simplify type signatures.
+
+The syntactic shortcut syntax is orthogonal to the rest of the proposal and could be entirely removed. The same functionality could potentially be achieved with QuasiQuoters.
+
+As an extension to the proposed behavior, local imports could be used to shadow globally-defined symbols. As an example, the ``blaze-html`` library provides symbols for ``head``, ``div`` and ``id`` ; for this reason,  the relevant modules are frequently imported qualified, or those symbols are explicitly hidden with ``-XNoImplicitPrelude`` and an explicit import. This is necessary for preventing uses of those symbols to be reported as ambiguous by the compiler. Without type-driven disambiguation, this is the only sane behavior in current Haskell, which only allows a single, unordered list of module imports ; however, local imports could be seen as defining nested scopes, such that:
```

Comment:
This also raises another question actually.

What does the following do:

```haskell
f = id

g = id
  where
    import Prelude hiding (id)
```

do?

(I'd imagine it doesn't hide `id` any further, it just doesn't import it a second time, but it should be specified more precisely in the specification section).